validator/src/main.rs

```diff
@@ -1,186 +1,185 @@
-        resp
+        data == token,
+        "expected client to receive {:?}; got {:?} instead",
+        String::from_utf8_lossy(&*token),
+        String::from_utf8_lossy(&*data),
     );
     Ok(())
 }
 
-#[tracing::instrument(name = "outbound_server", skip(ready, shutdown))]
-async fn outbound_serve(
-    listen_addr: SocketAddr,
-    ready: Arc<Notify>,
-    shutdown: kubert::shutdown::Watch,
-    resp: String,
-) -> Result<()> {
-    let listener = TcpListener::bind(listen_addr)
-        .await
-        .expect("Failed to bind server");
-    info!("Listening for incoming connections");
-
-    ready.notify_one();
-
-    tokio::select! {
-        _ = accept(listener, resp).in_current_span() => unreachable!("`accept` function never returns"),
-        _ = shutdown.signaled() => debug!("Received shutdown signal")
-    }
-
-    Ok(())
-}
-
-async fn accept(listener: TcpListener, resp: String) {
+/// Accepts connections from `listener`, writes `token` to the socket, and then closes the
+/// connection.
+#[tracing::instrument(level = "debug", skip_all)]
+async fn serve(listener: TcpListener, token: Bytes) {
     loop {
-        let (mut stream, client_addr) = listener
+        let (mut socket, client_addr) = listener
             .accept()
             .await
             .expect("Failed to establish connection");
-        info!("Accepted connection");
-        let rng_resp = resp.clone();
-        let _ = tokio::spawn(
+        let token = token.clone();
+        tokio::spawn(
             async move {
-                let resp_bytes = rng_resp.as_bytes();
-                // We expect this write to complete instantaneously,
-                // a timeout is not needed here.
-                match stream.write_all(resp_bytes).await {
-                    Ok(()) => debug!(written_bytes = resp_bytes.len()),
+                debug!("Accepted");
+                // We expect this write to complete instantaneously, so a timeout is not needed
+                // here.
+                match socket.write_all(&*token).await {
+                    Ok(()) => debug!(bytes = token.len(), "Wrote message to client"),
                     Err(error) => error!(%error, "Failed to write bytes to client"),
                 }
             }
-            .instrument(tracing::info_span!("conn", %client_addr)),
+            .instrument(tracing::info_span!("conn", client.addr = %client_addr)),
         );
     }
 }
 
+/// Connects to the target address and reads exactly `size` bytes.
+#[tracing::instrument(level = "debug", skip_all)]
+async fn connect(addr: SocketAddr, size: usize) -> Result<Bytes> {
+    let mut socket = TcpStream::connect(addr).await?;
+    debug!(client.addr = %socket.local_addr()?, "Connected");
+
+    socket
+        .readable()
+        .await
+        .context("cannot read from client socket")?;
+
+    let mut buf = BytesMut::with_capacity(size);
+    while buf.len() != size {
+        let size = socket.read_buf(&mut buf).await?;
```

Comment:
Will shadowing `size` here have an effect on the outer loop condition? Could this like finish the loop in one iteration if for example we only partially fill the buffer in the first read?

Reply:
Practically there are no issues here, but it may be confusing enough to a reader that we might want to call this variable `sz`...

> Could this like finish the loop in one iteration if for example we only partially fill the buffer in the first read?

If we only partially fill the buffer, read_buf will return the number of bytes read and `size` will reflect that value.

The outer `size` is **only** used as the capacity for the buffer in this code.